### PR TITLE
Prevent 500 on parallel object creation

### PR DESCRIPTION
**Pull request by kathap**
While fixing #3899, we found further possible occurrences of 500 errors in severel _create actions when doing parallel requests. In this PR we want to address them by adding an around_save to the corresponding model in which we catch Sequel::UniqueConstraintViolation errors for the case when  ne POST bypasses the other one and inserts the record into the db after the first POST has passed the validate function but before the actual insert.

* A short explanation of the proposed change:
Adding an around_save to the corresponding model in which we catch Sequel::UniqueConstraintViolation errors for the case when  ne POST bypasses the other one and inserts the record into the db after the first POST has passed the validate function but before the actual insert
* An explanation of the use cases your change solves
No more 500 errors on parallel request execution

* Links to any other associated PRs
#3899
#3924
* [X] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [X] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/spec/README.md#cf-acceptance-tests-cats)

### CODE REVIEW

File: spec/unit/actions/service_credential_binding_app_create_spec.rb [hunk 1]
```diff
@@ -145,6 +145,57 @@ module V3
...
+              end.not_to raise_error
+
+              # Mock the validation for the second request to simulate the race condition and trigger a
+              # unique constraint violation on service_instance_guid + app_guid
+              allow_any_instance_of(ServiceCredentialBindingAppCreate).to receive(:validate_binding!).and_return(true)
+              allow_any_instance_of(ServiceBinding).to receive(:validate).and_return(true)
+              binding = instance_double(ServiceBinding, destroy: false)
+              allow(ServiceBinding).to receive(:first).with(service_instance:, app:).and_return(binding)
+
+              allow_any_instance_of(ServiceBinding).to receive(:save_with_attributes_and_new_operation) do |instance|
```
Comment (svkrieger) [inline spec/unit/actions/service_credential_binding_app_create_spec.rb:187]: Same question here

File: spec/unit/isolation_segment_create_spec.rb [hunk 2]
```diff
@@ -44,6 +44,24 @@ module VCAP::CloudController
...
+            IsolationSegmentCreate.create(message)
+          end.not_to raise_error
+
+          # Mock the validation for the second request to simulate the race condition and trigger a unique constraint violation
+          allow_any_instance_of(IsolationSegmentModel).to receive(:validate).and_return(true)
+
+          # Second request, should fail with correct error
+          expect do
+            IsolationSegmentCreate.create(message)
+          end.to raise_error(IsolationSegmentCreate::Error)
```
Comment (svkrieger) [inline spec/unit/isolation_segment_create_spec.rb:62]: Would be good to test for the error string here, like done in all other tests.

**Review by svkrieger — COMMENTED** [review]
Looks great! Just a question about mocking in two tests and a minor finding.

EDIT: Also do we plan to get rid of the unique checks in the coding? If we catch the unique constraint errors thrown by the DB reliably now, it should not be necessary to check for uniqueness beforehand.

File: spec/unit/actions/service_credential_binding_app_create_spec.rb [hunk 3]
```diff
@@ -145,6 +145,57 @@ module V3
...
+              end.not_to raise_error
+
+              # Mock the validation for the second request to simulate the race condition and trigger a
+              # unique constraint violation on app_guid + name
+              allow_any_instance_of(ServiceCredentialBindingAppCreate).to receive(:validate_binding!).and_return(true)
+              allow_any_instance_of(ServiceBinding).to receive(:validate).and_return(true)
+              binding = instance_double(ServiceBinding, destroy: false)
+              allow(ServiceBinding).to receive(:first).with(service_instance:, app:).and_return(binding)
+
+              allow_any_instance_of(ServiceBinding).to receive(:save_with_attributes_and_new_operation) do |instance|
```
[the remark this answers is not in the document]
Comment (kathap) [reply spec/unit/actions/service_credential_binding_app_create_spec.rb]: Good point!

File: spec/unit/actions/service_credential_binding_key_create_spec.rb [hunk 4]
```diff
@@ -250,6 +250,34 @@ module V3
             end
           end
         end
+
+        context 'when creating keys with the same name concurrently' do
+          let(:service_instance2) { ManagedServiceInstance.make(space:) }
```
Comment (svkrieger) [inline spec/unit/actions/service_credential_binding_key_create_spec.rb:255]: We don't need a second service instance, name and message here. The constraint is on the name of the keys per service instance.

**Review by svkrieger — CHANGES_REQUESTED** [review]

**Review by svkrieger — APPROVED** [review]

**Pull request merged**